### DISCUSSION TITLE
Hooking up to Jira and Confluence

### DISCUSSION THREAD

**Discussion by johan-lejdung** [khoj-ai/khoj#381] [Ideas]
Hooking Khoj up to Jira and Confluence would make it really powerful! 

It should be possible since both support API and Apps.

Lots of companies have organisation notes and files in Confluence and lots of discussions around active development take place on Jira.

**Comment by debanjum** [comment]
Hi @johan-lejdung, thanks for sharing the suggestion! Given our limited resources we may not be able to pursue this immediately. But Khoj has (beta) support for plugins. So you could write a script to convert Jira or Confluence content into a standardized jsonl format for Khoj to index for search and chat. I can provide guidance if you (or anyone) wants to build this out?

**Reply by technicaldirector** [reply]
Hey @debanjum, I'm currently looking into that. I've written a Python script that converts the data into JSONL format. Could you clarify if the JSONL needs to follow a specific structure? Also, what's the recommended way to upload this to the Khoj knowledge base? I noticed JSON files can't be uploaded directly. Does this mean it only works via the Khoj API?

**Reply by technicaldirector** [reply]
`{"title": "Title", "content": "This is the content", "source": "confluence", "url": "http://wiki.code/display/AN/Title"}`

That is my current structure

**Reply by technicaldirector** [reply]
nevermind - just checked the code of the obsidian extension :)

**Reply by debanjum** [reply]
Yeah, using jsonl was the older way to index data. You should just push the data to index via the API. See the Swagger docs at `<your-khoj-url>/docs` (e.g http://localhost:42110/docs) or the Redocs at https://app.khoj.dev/redoc for API structure. 

And looking at the Khoj Obsidian client is good to see a working implementation. Glad you figured that out. Let me know if you hit any blockers indexing your confluence data.

Are you trying to index data to the Khoj cloud or a self-hosted instance? And what's the use-case like what kind of tasks/questions do you expect Khoj to help answer from your confluence data?

**Reply by technicaldirector** [reply]
Thanks for sharing the API structure link! I hadn't come across that URL anywhere else! 

At the moment, I’ve just been syncing to the cloud, but with the long-term plan of eventually migrating the data to a self-hosted instance. Due to our university’s internal structures and data security policies, a self-hosted setup is essential. 

The idea I had in mind was to sync the content from our university's confluence wiki to Khoj, so that students could simply query Khoj for information instead of digging through the wiki... Which is quite large and, honestly, hard to navigate 😅 

However, I’m not sure if that use case is currently supported. Please correct me if I’m wrong, but as far as I understand, documents can only be added on a per-user basis, and not shared across all users. Ideally, I’d like to sync all of our Confluence data under a root or super-admin account, so that this shared dataset could serve as a foundation for user-specific fine-tuning. 

Do you have a workaround in mind I could try for that?

**Reply by technicaldirector** [reply]
<img width="1912" height="451" alt="image" src="https://github.com/user-attachments/assets/be533dcd-512b-4852-8c40-f62e6089dd3e" />
<img width="1884" height="977" alt="image" src="https://github.com/user-attachments/assets/db21a642-7dab-4cd3-8e34-16f3a33a837d" />

I already tried to unset the default user - in hope that will make the doc available to all users but it doesn't seem to allow it.
Oh and while i am already asking. Haha feeling a bit dump but in the admin panel... Is there an edit button for the data i already added in a section? Haha wasn't able to find it. Currently used as workaround the url '/1/change/' to change already added stuff.

**Reply by debanjum** [reply]
Thought I'd already replied! The only way to share a knowledge base with a group of users is to create a protected agent (they can be shared via share link) and attach the knowledge base to the agent. 

This is a little clunky to share large knowledge bases but that's the best way currently. The more proper implementation would require adding the concept of groups to Khoj and sharing knowledge bases with the group.

**Reply by technicaldirector** [reply]
No worries at all, and thanks so much for taking the time to answer! 😊

Just to confirm my understanding: when self-hosting, making an agent public only makes it accessible to users on my own Khoj instance, right? It doesn’t share anything with any remote Khoj servers, correct?

I’ll dig into agents and see how they work for my use case. If they don’t fully meet my expectations, I’ll circle back to you and might look into contributing (will definitely check in with you first before doing so)
